packages/nextjs/test/utils/withSentry.test.ts

```diff
@@ -0,0 +1,92 @@
+import * as Sentry from '@sentry/node';
+import * as utils from '@sentry/utils';
+import { NextApiHandler, NextApiRequest, NextApiResponse } from 'next';
+
+import { AugmentedNextApiResponse, withSentry, WrappedNextApiHandler } from '../../src/utils/withSentry';
+
+const FLUSH_DURATION = 200;
+
+async function sleep(ms: number): Promise<void> {
+  await new Promise(resolve => setTimeout(resolve, ms));
+}
+
+async function callWrappedHandler(wrappedHandler: WrappedNextApiHandler, req: NextApiRequest, res: NextApiResponse) {
+  await wrappedHandler(req, res);
+
+  // Within the wrapped handler, we await `flush()` inside `res.end()`, but nothing awaits `res.end()`, because in its
+  // original version, it's sync (unlike the function we wrap around it). The original does actually *act* like an async
+  // function being awaited - subsequent steps in the request/response lifecycle don't happen until it emits a
+  // `prefinished` event (which is why it's safe for us to make the switch). But here in tests, there's nothing past
+  // `res.end()`, so we have to manually wait for it to be done.
+  await sleep(FLUSH_DURATION);
+  while (!res.finished) {
+    await sleep(100);
+  }
+}
+
+// We mock `captureException` as a no-op because under normal circumstances it is an un-awaited effectively-async
+// function which might or might not finish before any given test ends, potentially leading jest to error out.
+const captureExceptionSpy = jest.spyOn(Sentry, 'captureException').mockImplementation(jest.fn());
+const loggerSpy = jest.spyOn(utils.logger, 'log');
+const flushSpy = jest.spyOn(Sentry, 'flush').mockImplementation(async () => {
+  // simulate the time it takes time to flush all events
+  await sleep(FLUSH_DURATION);
+  return true;
+});
+
+describe('withSentry', () => {
+  let req: NextApiRequest, res: NextApiResponse;
+
+  const error = new Error('Oh, no! Charlie ate the flip-flops! :-(');
+
+  const origHandlerNoError: NextApiHandler = async (_req, res) => {
+    res.send('Good dog, Maisey!');
+  };
+  const origHandlerWithError: NextApiHandler = async (_req, _res) => {
+    throw error;
+  };
+
+  const wrappedHandlerNoError = withSentry(origHandlerNoError);
+  const wrappedHandlerWithError = withSentry(origHandlerWithError);
+
+  beforeEach(() => {
+    req = { url: 'http://dogs.are.great' } as NextApiRequest;
+    res = ({
+      send: function(this: AugmentedNextApiResponse) {
+        this.end();
+      },
+      end: function(this: AugmentedNextApiResponse) {
+        this.finished = true;
+      },
+    } as unknown) as AugmentedNextApiResponse;
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe('flushing', () => {
+    it('flushes events before rethrowing error', async () => {
```

Comment:
I don't think this title is very accurate. The test checks we rethrow an error and flushing is called, but there's nothing about the order.

Reply:
The order is implied, is it not? The second you throw an error, you're done, so as long as it both flushes and throws (which is what's checked in the test), you know it came in that order.

Reply:
If firstly `flush` is called, and then an exception is thrown, would the test pass? Whether it makes sense that `flush` is called before throwing the exception should be another reason to test it the test.

Let's say that a change introduces a call to `flush` early in the flow, and also a mistake that makes the last `flush` (the one currently present) not to be called. The test would still pass, but the behavior is not correct. In this case, the test should fail.

Reply:
Again, I think this is overkill. Tests are never going to prevent every possible bad refactor from getting through (because, among other things, sometimes refactors mean the tests themselves become wrong). Here and below, the big idea is that we had a problem: events weren't getting flushed in time, because flush either didn't get called or didn't have a chance to finish before the request ended. Both of these tests are there to prove that that particular problem isn't a problem anymore, by showing that flush _does_ get called, and _does_ have a chance to finish, which is exactly what they do.

I'm going to resolve both of these conversations because I don't think the naming of a test should block this fix getting out to users.